Convert `q` command to newshell

officialcjunior opened the pull request:
<!-- Filling this template is mandatory -->

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**
Initial attempt to convert `q` commands to newshell.
Created respective handlers for the existing commands and handled the quitting with preference case with subcommands.


**Test plan**

~~- [x] Firstly, we can fix handling `rz_cmd_quit_choose_hist_proj` (need to fix the names, too), 
I don't think arguments are handled properly there.~~

- [x] Split `q[y\n][y\n]` into multiple commands and handlers 

- [x] Pass arguments to `cmd_Quit` accordingly.

**Closing issues**

None

officialcjunior committed:
Initall attempt to convert `q` command to newshell

ret2libc commented on librz/core/cmd_descs.yaml:
```diff
@@ -590,9 +590,32 @@
+      cname: cmd_force_quit_without_history
+      summary: force quit rizin without saving history
+      args: []
+    - name: q
```
Maybe instead of this we can change a bit the way this thing is done... 
What if we just keep one `q` command that optionally accept arguments? First argument is "kill process", second argument is "save project". So if you want to kill and save project you would do `q y y`. If we go this route, just remove this and add the optional arguments in https://github.com/rizinorg/rizin/pull/525/files#diff-4f94ed53f57e9c49ad8a154a2d4e3defb758c0deaf69909274e560bb29a63938R598

@XVilka WDYT?

XVilka replied:
While it makes sense this is not how Vim commands work (Rizin commands are largely inspired by Vim).

ret2libc replied:
So you would stay with `qyy` `qyn` `qny` `qnn`?

XVilka replied:
I think `qyy` is fine probably. Not 100% sure.

ret2libc replied:
Ok, in this case @officialcjunior you should change this to be similar to https://github.com/rizinorg/rizin/blob/dev/librz/core/cmd_descs.yaml#L884

ret2libc commented on librz/core/cmd_descs.yaml:
```diff
@@ -590,9 +590,32 @@
+    - name: q
+      cname: cmd_quit_choose_hist_proj
+      summary: Quit rizin and choose to save projects and kill the process
+      args_str: " [y/n][y/n]"
```
```suggestion
      args_str: "<y|n><y|n>"
```

ret2libc commented on librz/core/cmd_quit.c:
```diff
@@ -32,6 +67,16 @@ static int cmd_Quit(void *data, const char *input) {
 }
 
+RZ_IPI RzCmdStatus rz_cmd_force_quit_handler(RzCore *core, int argc, const char **argv) {
+	cmd_Quit(core, *argv);
```
```suggestion
	cmd_Quit(core, argv[0] + 1);
```
The first character of argv[0] should be skip.

ret2libc commented on librz/core/cmd_quit.c:
```diff
@@ -15,6 +15,41 @@ static const char *help_msg_q[] = {
+	return RZ_CMD_STATUS_EXIT;
+}
+
+RZ_IPI RzCmdStatus rz_cmd_quit_choose_hist_proj (RzCore *core, int argc, const char **argv) {
```
this is not right. If we go down the route of having `qyn`/`qyy`/`qnn`/`qny`, the choose_project/choose_process are not separate arguments but you have to get those from the name of the command. You can just use different handlers in each case. Please see https://github.com/rizinorg/rizin/pull/525/files#r572686967 first.

officialcjunior replied:
Yes, now I have understood that. And yes, we will need to use different handlers.
👍🏾

officialcjunior committed:
Suggestions from code review: fix args_str

Co-authored-by: Riccardo Schirone <ret2libc@users.noreply.github.com>

officialcjunior committed:
Introduce `qyn/qyy/qnn/qny` and different handlers for each

Added individual handlers to set up `qyn`, `qyy`, `qnn` and
`qny` to preserve the vim-like bindings

ret2libc commented on librz/core/cmd_quit.c:
```diff
@@ -15,6 +15,34 @@ static const char *help_msg_q[] = {
+}
+
+RZ_IPI RzCmdStatus rz_quit_kill_save_handler(RzCore *core, int argc, const char **argv) {
+	core->num->value = 5;
```
Use some defines for these values (which I'm not sure what they mean... but let's keep them for now).
For example, RZ_QUIT_VALUE_KILL, RZ_QUIT_VALUE_SAVE, RZ_QUIT_VALUE_NOKILL, etc.

ret2libc commented on librz/core/cmd_quit.c:
```diff
@@ -15,6 +15,34 @@ static const char *help_msg_q[] = {
 
+RZ_IPI RzCmdStatus rz_cmd_quit_handler(RzCore *core, int argc, const char **argv) {
+	core->num->value = 0LL;
+	//exit (*input?rz_num_math (core->num, input+1):0);
```
Remove commented code please.

officialcjunior committed:
Use defines for the values used while quitting and removed code comments

Introduce `RZ_QUIT_VALUE_KILL` and three other defines to use when
handling the quit

officialcjunior committed:
Use clang-format on cmd_quit.c

ret2libc commented on librz/core/cmd_descs.yaml:
```diff
@@ -590,9 +590,42 @@
+    - name: "q"
+      cname: quit_with_preference
+      summary: Quit rizin and choose to save projects and kill the process
+      args_str: "<y|n><y|n>"
```
```suggestion
      options: "<y|n><y|n>"
```

ret2libc commented on librz/core/cmd_descs.yaml:
```diff
@@ -590,9 +590,42 @@
+      args_str: "<y|n><y|n>"
+      type: RZ_CMD_DESC_TYPE_INNER
+      subcommands:
+        - name: qyy
```
Do similar to `w1` / `w2`, etc.... Group `qyy` and `qyn` under `qy` and `qny`/`qnn` under `qn`.

officialcjunior committed:
Put `args_str:` for `q` as `options:`

Co-authored-by: Riccardo Schirone <ret2libc@users.noreply.github.com>

officialcjunior committed:
Group `qyy` and `qyn` under `qy` and `qny` and `qnn` under `qn`

ret2libc commented on librz/core/cmd_descs.yaml:
```diff
@@ -590,9 +590,48 @@
+    - name: "q"
+      cname: quit_with_preference
+      summary: Quit rizin and choose to save projects and kill the process
+      options: "<y|n><y|n>"
```
actually... sorry, for consistency this should be:
```suggestion
      options: "<yn><yn>"
```

ret2libc reviewed (commented):
LGTM, just fix that last thing and we can merge IMO.

officialcjunior committed:
Tweak options of `quit_with_preference` again

Co-authored-by: Riccardo Schirone <ret2libc@users.noreply.github.com>

officialcjunior committed:
Update cmd_descs.c according to the updated options for args

ret2libc commented on librz/core/cmd_descs.yaml:
```diff
@@ -590,9 +590,48 @@
+      args: []
+    - name: "q"
+      cname: quit_with_preference
+      summary: Quit rizin and choose to save projects and kill the process
```
```suggestion
      summary: Quit rizin and choose to kill the process and save projects
```
Because the first <yn> refers to the kill, the second to the projects.

officialcjunior replied:
Oh, yeah, missed that one. All of the summaries seem to be fine.

officialcjunior committed:
Change the overall summary of `q` according to its arguments

Co-authored-by: Riccardo Schirone <ret2libc@users.noreply.github.com>

officialcjunior committed:
Update cmd_descs.c according to the updated summary

ret2libc reviewed (approved):
Thanks!

ret2libc committed:
Merge branch 'dev' into q-newshell
